### PR TITLE
ocamlPackages.irmin: init at 1.4.0

### PR DESCRIPTION
**Pull request by alexfmpe**
###### Motivation for this change
Dependency of [tezos](https://gitlab.com/tezos/tezos)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by yorickvP** [comment]
Hi there! Are you packaging tezos? We did this work a while ago. Managed to produce some static binaries for it, too. [Here's the set of overrides we used to get it to build](https://gist.github.com/yorickvP/c273e44ce52d37eb3c8de196e9f8c438)

**Comment by yorickvP** [comment]
Also, feel free to cherry-pick some of https://github.com/serokell/nixpkgs/commits/ocaml-cross-fixes

**Comment by alexfmpe** [comment]
> Hi there! Are you packaging tezos? We did this work a while ago. Managed to produce some static binaries for it, too. Here's the set of overrides we used to get it to build

Yeah, kind of. I did something similar to that gist to get `dune build` to work at the root (using mainnet so far), and thought I might as well upstream the derivations. Yours seem slightly more up-to-date though, I haven't needed `ipaddr 4.0.0` for instance (though the dune file does state >= 4.0)

> Also, feel free to cherry-pick some of https://github.com/serokell/nixpkgs/commits/ocaml-cross-fixes

Oh thanks

File: pkgs/development/ocaml-modules/irmin/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+, astring, cstruct, fmt, hex, jsonm, logs, ocaml_lwt, ocamlgraph, uri
+}:
+
+buildDunePackage rec {
+  pname = "irmin";
+  version = "1.4.0";
+
+  minimumOCamlVersion = "4.03";
+
+  src = fetchFromGitHub {
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/default.nix]: The released sources from <https://github.com/mirage/irmin/releases/download/1.4.0/irmin-1.4.0.tbz> have correct version information.

File: pkgs/development/ocaml-modules/irmin/default.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  buildInputs = [ astring cstruct fmt hex jsonm logs ocaml_lwt ocamlgraph uri ];
+
+  doCheck = true;
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/mirage/irmin;
+    description = "Irmin, a distributed database that follows the same design principles as Git";
+    license = licenses.isc;
+    inherit (ocaml.meta) platforms;
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/default.nix]: This line is not needed: `buildDunePackage` sets the `meta.platforms` attribute to that value.

File: pkgs/development/ocaml-modules/irmin/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  minimumOCamlVersion = "4.03";
+
+  src = fetchFromGitHub {
+    owner = "mirage";
+    repo = pname;
+    rev = version;
+    sha256 = "0f272h9d0hs0wn5m30348wx7vz7524yk40wx5lx895vv3r3p7q7c";
+  };
+
+  buildInputs = [ astring cstruct fmt hex jsonm logs ocaml_lwt ocamlgraph uri ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/default.nix]:
```suggestion
  propagatedBuildInputs = [ astring cstruct fmt hex jsonm logs ocaml_lwt ocamlgraph uri ];
```

File: pkgs/development/ocaml-modules/irmin/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{ stdenv, fetchFromGitHub, buildDunePackage, ocaml
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/default.nix]: You don’t need `stdenv` as argument when you only use `stdenv.lib`: you can directly take `lib` as argument.

**Review by vbgl — CHANGES_REQUESTED** [review]

**Comment by yorickvP** [comment]
@alexfmpe I was also just told about https://github.com/serokell/tezos-client-packaging, which produces static tezos client binaries

File: pkgs/top-level/ocaml-packages.nix [hunk 5]
```diff
@@ -314,6 +314,8 @@ let
       then callPackage ../development/ocaml-modules/ipaddr { }
       else ipaddr_p4;
 
+    irmin = callPackage ../development/ocaml-modules/irmin { };
```
Comment (vbgl) [inline pkgs/top-level/ocaml-packages.nix]:
```suggestion
    irmin_1 = callPackage ../development/ocaml-modules/irmin/1.4.nix { };
```

Can you please leave room for irmin 2? Thanks.

File: pkgs/development/ocaml-modules/irmin/default.nix (hunk 1 above)
Comment (alexfmpe) [reply pkgs/development/ocaml-modules/irmin/default.nix]: Meaning I should directly use `fetchurl` ?

Comment (vbgl) [reply pkgs/development/ocaml-modules/irmin/default.nix]: Yes.

**Review by vbgl — APPROVED** [review]

**Pull request merged**